### PR TITLE
Adding EDN transport, take 2

### LINKED ISSUE #60: Implement an EDN transport
Having an EDN transport would simplify the lives of people using Clojure/ClojureScript/Java clients for nREPL. It should be pretty easy to create a transport based on the default bencode transport that simply uses EDN instead of bencode. Here's the implementation of the `bencode` transport as a reference:

```clj
(defn bencode
  "Returns a Transport implementation that serializes messages
   over the given Socket or InputStream/OutputStream using bencode."
  ([^Socket s] (bencode s s s))
  ([in out & [^Socket s]]
   (let [in (PushbackInputStream. (io/input-stream in))
         out (io/output-stream out)]
     (fn-transport
      #(let [payload (rethrow-on-disconnection s (bencode/read-bencode in))
             unencoded (<bytes (payload "-unencoded"))
             to-decode (apply dissoc payload "-unencoded" unencoded)]
         (merge (dissoc payload "-unencoded")
                (when unencoded {"-unencoded" unencoded})
                (<bytes to-decode)))
      #(rethrow-on-disconnection s
                                 (locking out
                                   (doto out
                                     (bencode/write-bencode %)
                                     .flush)))
      (fn []
        (if s
          (.close s)
          (do
            (.close in)
            (.close out))))))))
```

The best thing about this transport is that it requires no extra deps, so it can be easily bundled with nREPL itself.

### PR DESCRIPTION
**Pull request by shen-tian**
My attempt to finish what @dotemacs started in #67. Closes #60 

I followed the discussion under the [previous PR](#67), which provided some great sign-posts. What I did, in rough order:

- Rebased @dotemacs's commits onto master,
- Pending the discussion that started in the last PR, and I'm sure will continue, around how EDN transport will work, I've taken a permissive approach of aggressively converting all keywords (keys or values) to strings within the transport for both send and receive. This may not be what we ultimately want to do, but allowed rest of the work to continue.
- Update: we can do without this now! ~~[Hacked around](https://github.com/nrepl/nrepl/commit/34ce2bc18f358be8b971dd78484bdabd7c4b0b6e) a "can't read UUID as a number" bug, which only happened for UUIDs started with a digit. The workaround is all our rand UUIDs start with a letter. This can't go into production, of course. Ideas? @pfeodrippe , you talked a bit about the bug [here](https://github.com/nrepl/nrepl/pull/67#issuecomment-445779570)~~
- Converted all the test to use a "normalized" form of `:status`, a set of keywords. The converting to sets in NB because the EDN transport preserves the set structure, whereas bencode did not. The keywordizing might not be as important, pending the previous question. This is mostly testing code though, so not a design decision.
- The biggest breakthrough was this [commit](https://github.com/nrepl/nrepl/commit/7057d0ceadf9a571d250acf4b814d8558ecf3135). I don't know why this works, but it fixed most of the printing related tests. `*print-readably*` is being left in an odd state by the print middleware, as far as I can gather. @cichli might have some insight?

There's still quite some work left to do, but I think the problem's back is broken. 

To finish this, from me:

- ~~Deciding how permissive/strict the EDN transport should be, and update accordingly.~~
- ~~Figuring out the underlying reason for the UUID reading bug, and remove the workaround.~~
- Figuring out the underlying reason for needing the `*print-readably*` binding.
- ~~Fixing the remaining test or two,~~ Update: fixed ~~and have a look at the assertion that was removed [here](https://github.com/nrepl/nrepl/commit/c17b8c43c66498562cf5d4f050769490b1c24538)~~
- ~~Some manual testing with a REPL implementation?~~ 

In the meanwhile, comments/ideas welcome.

### Checklist

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [x] You've added tests to cover your change(s)
- [x] All tests are passing
- [x] The new code is not generating reflection warnings
- [x] You've updated the [changelog](../CHANGELOG.md)(that only applies to user-visible changes)

### CODE REVIEW

**Comment by arichiardi** [comment]
Wow awesome job, thank you! Following from far🍿

**Comment by shen-tian** [comment]
Managed to fix the last consistently breaking test! Gonna make a list of things that sporadically fail to squish:

- [ ] `edn-transport-communication - nrepl.edn-test` Something to do with not closing a socket/releasing the port properly? [example](https://circleci.com/gh/shen-tian/nrepl/475)
- [ ] `interrupt-load-file - nrepl.core-test` [example](https://circleci.com/gh/shen-tian/nrepl/473)

Fails but not in scope

- `streamed-printing - nrepl.core-test` - As reported in #132

**Comment by bbatsov** [comment]
Great progress! I'll review the code a bit later.

Are things behaving well when you test them manually in a REPL?

File: src/clojure/nrepl/core.clj [hunk 1]
```diff
@@ -219,6 +221,8 @@
                            (merge connect-defaults
                                   (socket-info uri))))))
 
+(add-socket-connect-method! "edn" {:transport-fn transport/nrepl+edn
```
Comment (bbatsov) [inline src/clojure/nrepl/core.clj]: I actually wanted us to name the protocol `nrepl+edn` and the function itself `edn`.

File: src/clojure/nrepl/core.clj [hunk 2]
```diff
@@ -66,7 +66,9 @@
 (defn- delimited-transport-seq
   [client termination-statuses delimited-slots]
   (with-meta
-    (comp (partial take-until (comp #(seq (clojure.set/intersection % termination-statuses))
+    (comp (partial take-until (comp #(seq (clojure.set/intersection
+                                           (keyworded-set %)
```
Comment (bbatsov) [inline src/clojure/nrepl/core.clj]: Why is this change needed?

File: src/clojure/nrepl/misc.clj [hunk 3]
```diff
@@ -60,3 +60,8 @@
         (require (symbol (namespace sym)))
         (resolve sym)
         (catch Exception _))))
+
+(defn keyworded-set
```
Comment (bbatsov) [inline src/clojure/nrepl/misc.clj]: Such helpers should be kept private and in the namespaces that need them. The misc ns exposes functions useful to client writes (the name is a bit confusing).

File: src/clojure/nrepl/transport.clj [hunk 4]
```diff
@@ -123,6 +130,40 @@
             (.close in)
             (.close out))))))))
 
+(defn stringify-everything
```
Comment (bbatsov) [inline src/clojure/nrepl/transport.clj]: Helpers should be private.

Comment (bbatsov) [reply src/clojure/nrepl/transport.clj]: Btw, this affects only keywords, right?

File: src/clojure/nrepl/transport.clj [hunk 5]
```diff
@@ -123,6 +130,40 @@
...
+
+(defn nrepl+edn
+  "Returns a Transport implementation that serializes messages
+   over the given Socket or InputStream/OutputStream using EDN."
+  ([^Socket s] (nrepl+edn s s s))
+  ([in out & [^Socket s]]
+   (let [in (java.io.PushbackReader. (io/reader in))
+         out (io/writer out)]
+     (fn-transport
+      #(rethrow-on-disconnection s (stringify-everything (edn/read in)))
```
Comment (bbatsov) [inline src/clojure/nrepl/transport.clj]: That was done mostly to appease the tests, right? In general it seems redundant to convert the keywords to strings.

**Comment by shen-tian** [comment]
@bbatsov I haven't tried yet. I am a bit worried about all the dynamic print related vars. Not sure how much this is just a feature of nREPL v.s. something odd going on elsewhere in the system. Will try that later. In the meanwhile:

## On the EDN transport protocol:

What do we accept, and what do we guarantee in sending back? The default solution is that we just stick to the implicit (because it's never been exposed?) internal protocol for both, e.g.

```clj
{:op :eval :code "(+ 2 2)"}
```
### Pros:

A major plus for EDN transport is to serve Clojure(Script) clients, which will all be able to speak EDN natively. Also, we can always be strict now, and relax things later, rather than become stricter and causing breaking changes.

### Cons:

Cons: This will take quite some documenting/checking, as it's more prescriptive than the simpler bencode transport, which is stringify-everything. The would extend to all the middleware too, as they'd all be able to speak EDN now. This is kind of like having everyone seeing colour for the first time... and immediately requiring colour coordinated uniforms. I'm not even sure this is possible without a coordinated effort with all middlewares, as in essence, we are requiring more detail in the specification of all of them. Overall, this might not be possible in practice? @cgrand mentioned Postel's law in #67. Maybe this is more applicable to third party middlewares than clients?

### Alternatives:

#### Stringy-fy everything:

Requires not much more specification than what already exists. We'd need to specify `:status` is either a string or a set of strings etc., but I'm not aware of any other EDN structures that we are now exposing from the internals, that was previously hidden behind the simplicity of bencode. 

The PR as it stands follows this, though more to facilitate discussion, (and having passing tests as a starting point) than it being my preference.

#### Permissive on input, strict on output

Think we'll still have the same problem as before, as some stuff that's pretty much implementation detail now might become the defacto spec, possibly without notice to middlware authors.

#### Strict on everything we know about, be relaxed on the rest?

Between `cider-nrepl` `refactor-nrepl` `piggieback` and the build-in middlewares make up the bulk of real life ones, we can just define what they do, and leave "third-party" middleware to deal with the specification problem?

**Comment by bbatsov** [comment]
The middlewares don't actually decide what type of data to return - that's entire up to the transport. In practice all middlewares kept to the simply data structures supported by bencode (namely lists and maps of primitive data types like strings and numbers) and we can't really change this, although we can have some middleware returning more complex data structures when they using the EDN transport. 

We will decide this down the road, but for starters I'd just replicate what we have right now. Bencode requests and responses couldn't contain keywords because bencode can't represent them, but with EDN that's obviously not a problem, so I'd just move the current stringification to the bencode transport implementation (as opposed to being global as it is now). Then every transport can individually decide what structure of requests and responses it requires. I doubt any client would aim to support different transports (as this is a bit pointless), so this differences won't really be felt most of the time. 

Does this make sense to you?

**Comment by bbatsov** [comment]
This is where we messed up in the past https://github.com/nrepl/nrepl/blob/master/src/clojure/nrepl/transport.clj#L40 It was marked to be fixed a long time ago, but with only one transport it wasn't that big of a deal anyways.

File: src/clojure/nrepl/core.clj (hunk 1 above)
Comment (shen-tian) [reply src/clojure/nrepl/core.clj]: Ok. So swap those two around? Will do.

File: src/clojure/nrepl/core.clj (hunk 2 above)
Comment (shen-tian) [reply src/clojure/nrepl/core.clj]: tbh, not sure. This is from one of @dotemacs's commits.

File: src/clojure/nrepl/misc.clj (hunk 3 above)
Comment (shen-tian) [reply src/clojure/nrepl/misc.clj]: I suspect they can be all moved into `core-test`, where it's mostly used.

File: src/clojure/nrepl/transport.clj (hunk 4 above)
Comment (shen-tian) [reply src/clojure/nrepl/transport.clj]: The last two comments will be impacted by the protocol story.

**Comment by shen-tian** [comment]
> The middlewares don't actually decide what type of data to return - that's entirely up to the transport. In practice all middlewares kept to the simply data structures supported by bencode (namely lists and maps of primitive data types like strings and numbers) and we can't really change this, although we can have some middleware returning more complex data structures when they using the EDN transport.

Technically true, yes, but the method you favour, if I understand, is for the EDN transport to be transparent? This would expose (previously unexposed) workings of the middleware to the outside. This was hidden by the limitation of bencode previously. This is, in practice, making this canonical. 

For example, I think as it stands, `{:op :eval :code "(+ 2 2)"}` is not actually valid with a transparent transport. The version that would work is `{:op "eval" :code "(+ 2 2)"}`. This was previously discussed [here](https://github.com/nrepl/nrepl/pull/67#issuecomment-445783776) but not resolved.

This contradicts the published [documents](https://nrepl.org/nrepl/0.6.0/ops.html) but didn't matter, because the bencode stripping and keywordizing at the end hid that detail before.

File: src/clojure/nrepl/core.clj (hunk 2 above)
Comment (bbatsov) [reply src/clojure/nrepl/core.clj]: Well, I think you should check if this actually breaks something as it looks weird to me.

File: src/clojure/nrepl/core.clj (hunk 1 above)
Comment (bbatsov) [reply src/clojure/nrepl/core.clj]: 👍

**Comment by bbatsov** [comment]
> For example, I think as it stands, {:op :eval :code "(+ 2 2)"} is not actually valid with a transparent transport. The version that would work is {:op "eval" :code "(+ 2 2)"}. This was previously discussed here but not resolved.

Yeah, this is a good point. Probably we should make the EDN transport permissive as well on the request level, but the conversion of keywords should be done only in the transport itself and we should rely on the top-level conversion the way the bencode transport does it. Btw, I do think that internally we should be converting everything to keywords, not the other way around.

**Comment by shen-tian** [comment]
I hear you on changing all the keywordize/stringify into transports. Should we tackles it in this PR?

We can be permissive on the request in side... but would still need to decide what we emit though. I can go at sketching out a (cloujure.)spec in spirit of what you are describing. (btw, how do you feel about having a spec for response + requests? useful?)

Do you think there's room to release this feature as an "alpha/beta/experimental" feature, i.e. might change the format a bit? I'm a bit weary of a yak shaving mission where we have to be way more sure of the format than we need to be.

**Comment by bbatsov** [comment]
> I hear you on changing all the keywordize/stringify into transports. Should we tackles it in this PR?

No, we can do this later.

> We can be permissive on the request in side... but would still need to decide what we emit though. I can go at sketching out a (cloujure.)spec in spirit of what you are describing. (btw, how do you feel about having a spec for response + requests? useful?)

That's a pretty cool idea! Specing the requests and responses would be super useful in the long run. 

> Do you think there's room to release this feature as an "alpha/beta/experimental" feature, i.e. might change the format a bit? I'm a bit weary of a yak shaving mission where we have to be way more sure of the format than we need to be.

Certainly. I don't want us to overdo things right away. I think that releasing something that works would be a great start and we can polish it later. Makes sense to stick to the current request/response format for the first release. Then a next step would to consider a simple mechanism to supply different responses based on the transport, but that's orthogonal to creating the transport.

File: src/clojure/nrepl/core.clj (hunk 2 above)
Comment (shen-tian) [reply src/clojure/nrepl/core.clj]: Reverted. Everything seems to still work.

**Comment by shen-tian** [comment]
Ok! Progress. Addressed all the comments I think, and we have some resolution on the question of the EDN protocol spec. It's the same as the internals, except I've put a shim in to make `:op` keywords, rather than strings. We can change the internals to actually use keywords... later. Once that's done, the transport can actually be transparent. I think this satisfies the need to get the transport out there. The shim ensures it doesn't contradict the docs, at least.

As long as we reserve the rights to adjust how the transport works later, I'm happy with that aspect for now.

In the course of this, I've moved the keywordizing/stringfication out of `FnTransport`. This might be a breaking change for anything that depended on it? But should be easy enough to fix at this stage. 

I'll do some manual testing of the transport next..

File: src/clojure/nrepl/misc.clj (hunk 3 above)
Comment (shen-tian) [reply src/clojure/nrepl/misc.clj]: Removed most usages of this. Then rolled into `clean-response` in `core-test`

File: test/clojure/nrepl/core_test.clj [hunk 6]
```diff
@@ -70,7 +72,8 @@
 
 (defmacro def-repl-test
   [name & body]
-  `(deftest ~(with-meta name (merge {:private true} (meta name)))
+  ;; TODO: decide what to do with the private tags before merging
+  `(deftest ~(with-meta name (merge #_{:private true} (meta name)))
```
Comment (shen-tian) [inline test/clojure/nrepl/core_test.clj]: Why do we make the repl tests private? It stops one from using test selectors like this from working:

```
lein with-profile +1.10,+test test :only nrepl.core-test/transports-fail-on-disconnects
```

I'd like to remove the private bit if possible. This is also the one line that's generating eastwood warnings.

Comment (bbatsov) [reply test/clojure/nrepl/core_test.clj]: No idea. Likely some oversight. Feel free to remove this.

**Comment by bbatsov** [comment]
Everything looks great to me at this point! I'll merge this once you're done with the manual testing.

**Comment by shen-tian** [comment]
Just checking we are talking the same thing:

Did some manual testing per https://nrepl.org/nrepl/0.6.0/hacking_on_nrepl.html

Used variations of this code after running `lein repl`

```clj
(require '[nrepl.server :refer [start-server stop-server]])
(require '[nrepl.core :as nrepl])

(def server (start-server :port 7889 :transport-fn nrepl.transport/edn))
(with-open [conn (nrepl/connect :port 7889 :transport-fn nrepl.transport/edn)]
           (let [resp (-> (nrepl/client conn 1000)
                          (nrepl/message {:op :eval :code "(+ 2 2)"}))]
             (->> resp
                  (map nrepl/read-response-value)
                  nrepl/combine-responses)))
```

Seems to work as expected.

File: test/clojure/nrepl/core_test.clj (hunk 6 above)
Comment (shen-tian) [reply test/clojure/nrepl/core_test.clj]: 👍

**Comment by bbatsov** [comment]
Yeah, that's exactly what I meant. Seems we're in business then! 

I see there's some linting failure and some failing test `nrepl.core-test/test-url-connect` (I assume this happens randomly). Perhaps some race condition in the tests or something like this?

**Comment by shen-tian** [comment]
I've made a list of tests that randomly fail above. I don't think they are EDN transport related, but rather what you see if you run through the test suite 250+ times :)

Would still like to fix those though... but later.

You are not worried about having to set the `*print-readably*` to make the conversion work? If not, I'm done on this one.

**Comment by bbatsov** [comment]
I had forgotten about this. Seems it's true by default, so I'm not worried about it https://clojuredocs.org/clojure.core/*print-readably*

**Comment by shen-tian** [comment]
> I had forgotten about this. Seems it's true by default, so I'm not worried about it https://clojuredocs.org/clojure.core/*print-readably*. Just add some todo in the code, and perhaps a ticket for @cichli to take a look into this problem when he has some time.

Left a note :)
